src/internal/ppsutil/decoder_test.go

```diff
@@ -0,0 +1,114 @@
+package ppsutil_test
+
+import (
+	"encoding/json"
+	"io"
+	"reflect"
+	"strings"
+	"testing"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/ppsutil"
+)
+
+func TestSpecReader(t *testing.T) {
+	var cases = map[string]struct {
+		input     string   // YAML
+		expected  []string // list of JSON specs
+		expectErr bool
+	}{
+		"empty input is okay": {},
+		"invalid JSON is bad": {
+			input:     `"dkjfskjf`,
+			expectErr: true,
+		},
+		"valid JSON is valid": {
+			input:    `{"pipeline": {"name": "test"}}`,
+			expected: []string{`{"pipeline": {"name": "test"}}`},
+		},
+		"valid spec is valid": {
+			input:    `{"pipeline": {"name": "test"}, "datumTries": 1, "resourceRequests":{"disk":"256Mi", "cpu": .5}, "autoscaling":true}`,
+			expected: []string{`{"pipeline": {"name": "test"},"resourceRequests":{"disk":"256Mi", "cpu": 0.5},"datumTries": 1, "autoscaling":true}`},
+		},
+		"invalid spec is invalid": {
+			input:     `{"not_aField": 2}`,
+			expectErr: true,
+		},
+		"multiple specs work": {
+			input: `{"pipeline": {"name": "test"}}
+---
+{"pipeline": {"name": "test2"}}`,
+			expected: []string{`{"pipeline": {"name": "test"}}`, `{"pipeline": {"name": "test2"}}`},
+		},
+		"lists work": {
+			input: `
+- {"pipeline": {"name": "test"}}
+- {"pipeline": {"name": "test2"}}`,
+			expected: []string{`{"pipeline": {"name": "test"}}`, `{"pipeline": {"name": "test2"}}`},
+		},
+		"null works": {
+			input: `
+                                  pipeline:
+                                    name: test
+                                  resourceRequests: null`,
+			expected: []string{`{"pipeline": {"name": "test"},"resourceRequests": null}`},
+		},
+		"sequences work": {
+			input: `
+                                 pipeline:
+                                   name: test
+                                 transform:
+                                   cmd: ["foo", "bar"]`,
+			expected: []string{`{"pipeline": {"name": "test"}, "transform": {"cmd": ["foo", "bar"]}}`},
+		},
+	}
+cases:
+	for n, c := range cases {
+		var (
+			r   = ppsutil.NewSpecReader(strings.NewReader(c.input))
+			got []string
+			i   int
+		)
+		for {
+			g, err := r.Next()
+			if errors.Is(err, io.EOF) {
+				break
+			}
+			if err != nil {
+				if !c.expectErr {
+					t.Errorf("%s[%d]: %v", n, i, err)
+				}
+				continue cases
+			}
+			got = append(got, g)
+			i++
+		}
+		if c.expectErr {
+			t.Errorf("%s: error expected; got %v", n, got)
+			continue
+		}
+		if len(got) != len(c.expected) {
```

Comment:
```go
if got, want := len(got), len(c.expected); got != want {
    t.Errorf("%s: len(expected):\n   got: %v\n want: %v", n, got, want)
}
```
https://github.com/golang/go/wiki/TestComments#got-before-want

The style guide mentions `if got, want := ...`: https://google.github.io/styleguide/go/decisions#data-driven-test-cases